src/client/dbus/thread_api_dbus.cpp

```diff
@@ -0,0 +1,115 @@
+ *    LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *    CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *    SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *    INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *    CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *    ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *    POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "client/dbus/thread_api_dbus.hpp"
+#include "common/code_utils.hpp"
+#include "dbus/constants.hpp"
+#include "dbus/dbus_message_helper.hpp"
+
+using namespace otbr::dbus;
+
+namespace otbr {
+namespace client {
+
+ThreadApiDBus::ThreadApiDBus(DBusConnection *aConnection)
+    : mInterfaceName("wpan0")
+    , mConnection(aConnection)
+{
+}
+
+ThreadApiDBus::ThreadApiDBus(DBusConnection *aConnection, const std::string &aInterfaceName)
+    : mInterfaceName(aInterfaceName)
+    , mConnection(aConnection)
+{
+}
+
+otError ThreadApiDBus::CallDBusMethodAsync(const std::string &aMethodName, DBusPendingCallNotifyFunction aFunction)
+{
+    otError          ret     = OT_ERROR_NONE;
+    DBusMessage *    message = nullptr;
+    DBusPendingCall *pending = nullptr;
+
+    message = dbus_message_new_method_call(dbus::kOtbrDbusPrefix.c_str(),
+                                           (dbus::kOtbrDbusObjectPrefix + mInterfaceName).c_str(),
+                                           dbus::kOtbrDbusPrefix.c_str(), aMethodName.c_str());
+    VerifyOrExit(message != nullptr, ret = OT_ERROR_FAILED);
+    VerifyOrExit(dbus_connection_send_with_reply(mConnection, message, &pending, DBUS_TIMEOUT_USE_DEFAULT) == true,
+                 ret = OT_ERROR_FAILED);
+
+    VerifyOrExit(dbus_pending_call_set_notify(pending, aFunction, this, &ThreadApiDBus::EmptyFree) == true,
+                 ret = OT_ERROR_FAILED);
+exit:
+    return ret;
+}
+
+otError ThreadApiDBus::Scan(const ScanHandler &aHandler)
+{
+    otError err = OT_ERROR_NONE;
+
+    VerifyOrExit(mScanHandler == nullptr, err = OT_ERROR_INVALID_STATE);
+    mScanHandler = aHandler;
+
+    err = CallDBusMethodAsync(kOtbrDbusObjectScanMethod, &ThreadApiDBus::sScanPendingCallHandler);
+    if (err != OT_ERROR_NONE)
+    {
+        mScanHandler = nullptr;
+    }
+exit:
+    return err;
+}
+
+void ThreadApiDBus::sScanPendingCallHandler(DBusPendingCall *aPending, void *aData)
+{
+    ThreadApiDBus *threadApiDBus = static_cast<ThreadApiDBus *>(aData);
+
+    threadApiDBus->ScanPendingCallHandler(aPending);
+}
+
+void ThreadApiDBus::ScanPendingCallHandler(DBusPendingCall *aPending)
+{
+    std::vector<otActiveScanResult> scanResults;
+    DBusMessage *                   message = dbus_pending_call_steal_reply(aPending);
+    auto                            args    = std::tie(scanResults);
+
+    if (message != nullptr)
+    {
+        DBusMessageToTuple(*message, args);
```

Comment:
If `DBusMessageToTuple` is always called using `std::tie` to receive results, I think it's better to change the `tuple` from reference to value and just call it like `DBusMessageToTuple(*message, std::tie(scanResults))`.